### PR TITLE
Add subgraph extraction to analytics library

### PR DESCRIPTION
**Pull request by gurbinder533**

### CODE REVIEW

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    katana::PropertyGraph* graph, const std::vector<uint32_t>& node_set) {
+  auto subgraph = std::make_unique<katana::PropertyGraph>();
+  if (node_set.empty()) {
+    return std::unique_ptr<katana::PropertyGraph>(std::move(subgraph));
+  }
+
+  uint64_t num_nodes = node_set.size();
+  // Subgraph topology : out indices
+  auto out_indices = std::make_unique<katana::LargeArray<uint64_t>>();
+  out_indices->allocateInterleaved(num_nodes);
```
Comment (uditagar) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:43]: i don't get this

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+  katana::gstl::Vector<katana::gstl::Vector<uint32_t>> subgraph_edges;
+  subgraph_edges.resize(num_nodes);
```
Comment (uditagar) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:46]: i also don;t get this

**Review by uditagar — COMMENTED** [review]
I have 2 small comments. Everything else looks fine. thank you

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp (hunk 1 above)
Comment (gurbinder533) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:43]: Can you be more specific, please?

Comment (uditagar) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:43]: I thought the size of indices can be more than num_nodes, no?

Comment (uditagar) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:43]: got this now; it's a prefix sum kind of thing

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp (hunk 2 above)
Comment (uditagar) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:46]: got this too. sorry about that.

**Review by uditagar — APPROVED** [review]

File: libgalois/include/katana/analytics/subgraph_extraction/subgraph_extraction.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+#define KATANA_LIBGALOIS_KATANA_ANALYTICS_SUBGRAPHEXTRACTION_SUBGRAPHEXTRACTION_H_
+
+#include "katana/PropertyGraph.h"
+#include "katana/analytics/Plan.h"
+
+// API
+
+namespace katana::analytics {
+
+/// A computational plan to for Total Triangle Counting.
```
Comment (roshandathathri) [inline libgalois/include/katana/analytics/subgraph_extraction/subgraph_extraction.h]: How is this related to TC?

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+ * including but not limited to those resulting from defects in Software and/or
+ * Documentation, or loss or inaccuracy of data of any kind.
+ */
+
+#include "katana/analytics/subgraph_extraction/subgraph_extraction.h"
+
+#include "katana/TypedPropertyGraph.h"
+#include "katana/analytics/Utils.h"
+
+using namespace katana::analytics;
```
Comment (roshandathathri) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: Move this to within `namespace {`

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    SubGraphExtractionPlan plan) {
+  if (auto r = katana::SortAllEdgesByDest(pg); !r) {
+    return r.error();
+  }
+
+  switch (plan.algorithm()) {
+  case SubGraphExtractionPlan::kNodeSet:
+    auto subgraph = SubGraphNodeSet(pg, node_set);
+    return std::move(subgraph.value());
+  }
```
Comment (roshandathathri) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:136]:
```cpp
default:
   return KATANA_ERROR(...);
}
```

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  return katana::ResultSuccess();
```
Comment (roshandathathri) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: should be error, not success.

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  katana::GAccumulator<uint64_t> accum_num_edges;
+  katana::do_all(
+      katana::iterate(node_set),
+      [&](const uint32_t& n) {
+        for (auto dest : node_set) {
+          auto edge_range = graph->topology().edge_range(n);
+          auto edge_matched = std::lower_bound(
+              edge_iterator(edge_range.first), edge_iterator(edge_range.second),
+              dest, [=](edge_iterator e, uint32_t n) {
+                return graph->topology().edge_dest(*e) < n;
```
Comment (roshandathathri) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: Why is this required? Add comments.

Also, are you using symmetric graph?

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+      katana::loopname("SubgraphExtraction"));
+
+  uint64_t num_edges = accum_num_edges.reduce();
+  // Subgraph topology : out dests
+  auto out_dests = std::make_unique<katana::LargeArray<uint32_t>>();
+  out_dests->allocateInterleaved(num_edges);
+
+  // Prefix sum
+  for (uint64_t i = 1; i < num_nodes; ++i) {
+    (*out_indices)[i] += (*out_indices)[i - 1];
```
Comment (roshandathathri) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:72]: You don't need `accum_num_edge`. `num_edges` should be `out_indices[num_nodes - 1]`

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+  out_indices->allocateInterleaved(num_nodes);
+
+  katana::gstl::Vector<katana::gstl::Vector<uint32_t>> subgraph_edges;
+  subgraph_edges.resize(num_nodes);
+
+  katana::GAccumulator<uint64_t> accum_num_edges;
+  katana::do_all(
+      katana::iterate(node_set),
+      [&](const uint32_t& n) {
+        for (auto dest : node_set) {
```
Comment (roshandathathri) [inline libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp:53]: This is fine for now. But an optimization could be to sort this set and then intersection this list with the neighbors list.

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp (hunk 7 above)
Comment (gurbinder533) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: Just doing binary search on the edges sorted by the dest id. No, the graph is not symmetric here.

Comment (roshandathathri) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: I got the binary search and sorted order part. My question was why do you need the lambda operator? You need to access the edge-dest? But in line 58, you seem to be doing it just by dereferencing the iterator?

Comment (gurbinder533) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: I need to iterate if there are multiple same edge-dst.

Comment (roshandathathri) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: That did not answer my question. I get the while loop. Please read my question again and check the code.

Comment (gurbinder533) [reply libgalois/src/analytics/subgraph_extraction/subgraph_extraction.cpp]: Got it. Thanks for pointing it out.

**Comment by gurbinder533** [comment]
Fixed another bug and added cmake test. Will merge once CI tests finish.
